File: _76_minimum_window_substring/main.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        if (t.size() > s.size())
        {
            return "";
        }

        // Much faster than classic hashmap
        int target[128] = {0};
        int current[128] = {0};
        int nb = 0;
        int bestLeft = 0;
        int bestLength = INT_MAX;
        int left = 0;
        int right = 0;
        
        // Fill target hashmap
        for (int i = 0; i < t.size(); i++)
        {
            target[t.at(i)]++;
        }

        while (left < s.size())
        {
            // Increment right until match or end
            while (nb != t.size() && right < s.size())
            {
                current[s.at(right)]++;
                if (current[s.at(right)] <= target[s.at(right)])
                {
                    nb++;
                }
                right++;
            }

            // Best match
            if (nb == t.size() && right - left < bestLength)
            {
                bestLeft = left;
                bestLength = right - left;
            }
            

            // Optimization
            current[s.at(left)]--;
            if (current[s.at(left)] < target[s.at(left)])
            {
                nb--;
                left++;
                continue;
            }

            left++;
        }
        
        if (bestLength == INT_MAX)
        {
            return "";
        }
        else
        {
            return s.substr(bestLeft, bestLength);
        }
    }
};

File: _76_minimum_window_substring/main.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if len(t) > len(s):
            return ""

        target = Counter(t)
        current = defaultdict(int)
        nb = 0
        left = 0
        right = 0
        bestLeft = 0
        bestLength = sys.maxsize

        while left < len(s):

            # Increment right until match or end
            while nb != len(t) and right < len(s):
                current[s[right]] += 1
                if current[s[right]] <= target[s[right]]:
                    nb += 1

                right += 1

            # Best match
            if nb == len(t) and right - left < bestLength:
                bestLeft = left
                bestLength = right - left

            # Decrement
            current[s[left]] -= 1
            if current[s[left]] < target[s[left]]:
                nb -= 1
                left += 1
                continue

            left += 1

        return "" if bestLength == sys.maxsize else s[bestLeft:bestLeft + bestLength]